File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-python/sdkwork_mail_sdk/provider_extension_catalog.py
from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class MailProviderExtensionCatalogEntry:
    extensionKey: str
    providerKey: str
    displayName: str
    surface: str
    access: str
    status: str


class MailProviderExtensionCatalog:
    recognizedSurfaces = [
        "control-plane",
        "runtime-bridge",
        "cross-surface",
    ]

    recognizedAccessModes = [
        "unwrap-only",
        "extension-object",
    ]

    recognizedStatuses = [
        "reference-baseline",
        "reserved",
    ]

    entries = [
        MailProviderExtensionCatalogEntry("smtp.transport", "smtp", "SMTP Transport", "runtime-bridge", "unwrap-only", "reserved"),
        MailProviderExtensionCatalogEntry("imap.sync", "imap", "IMAP Sync", "runtime-bridge", "unwrap-only", "reserved"),
    ]


def get_mail_provider_extension_catalog() -> list[MailProviderExtensionCatalogEntry]:
    return MailProviderExtensionCatalog.entries


def get_mail_provider_extension_descriptor(
    extension_key: str,
) -> Optional[MailProviderExtensionCatalogEntry]:
    for entry in MailProviderExtensionCatalog.entries:
        if entry.extensionKey == extension_key:
            return entry

    return None


def get_mail_provider_extensions_for_provider(
    provider_key: str,
) -> list[MailProviderExtensionCatalogEntry]:
    return [
        entry for entry in MailProviderExtensionCatalog.entries if entry.providerKey == provider_key
    ]


def get_mail_provider_extensions(
    extension_keys: list[str],
) -> list[MailProviderExtensionCatalogEntry]:
    entries: list[MailProviderExtensionCatalogEntry] = []
    for extension_key in extension_keys:
        entry = get_mail_provider_extension_descriptor(extension_key)
        if entry is not None:
            entries.append(entry)

    return entries


def has_mail_provider_extension(extension_keys: list[str], extension_key: str) -> bool:
    return extension_key in extension_keys and get_mail_provider_extension_descriptor(extension_key) is not None

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-typescript/src/provider-extension-catalog.ts
import { freezeMailRuntimeValue } from './runtime-freeze.js';
import type { MailProviderExtensionDescriptor } from './types.js';

export const mail_PROVIDER_EXTENSION_KEYS = freezeMailRuntimeValue(['smtp.transport', 'imap.sync'] as const);
export type MailKnownProviderExtensionKey = (typeof mail_PROVIDER_EXTENSION_KEYS)[number];

export const SMTP_TRANSPORT_mail_PROVIDER_EXTENSION_DESCRIPTOR: MailProviderExtensionDescriptor<MailKnownProviderExtensionKey> = freezeMailRuntimeValue({
  extensionKey: 'smtp.transport',
  providerKey: 'smtp',
  displayName: 'SMTP Transport',
  surface: 'runtime-bridge',
  access: 'unwrap-only',
  status: 'reserved',
});

export const IMAP_SYNC_mail_PROVIDER_EXTENSION_DESCRIPTOR: MailProviderExtensionDescriptor<MailKnownProviderExtensionKey> = freezeMailRuntimeValue({
  extensionKey: 'imap.sync',
  providerKey: 'imap',
  displayName: 'IMAP Sync',
  surface: 'runtime-bridge',
  access: 'unwrap-only',
  status: 'reserved',
});

export const mail_PROVIDER_EXTENSION_CATALOG: readonly MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[] = freezeMailRuntimeValue([
  SMTP_TRANSPORT_mail_PROVIDER_EXTENSION_DESCRIPTOR,
  IMAP_SYNC_mail_PROVIDER_EXTENSION_DESCRIPTOR
]);

const mail_PROVIDER_EXTENSION_DESCRIPTOR_BY_KEY = new Map<
  string,
  MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>
>(mail_PROVIDER_EXTENSION_CATALOG.map((descriptor) => [descriptor.extensionKey, descriptor]));

const EMPTY_mail_PROVIDER_EXTENSIONS = freezeMailRuntimeValue([] as const);

const mail_PROVIDER_EXTENSIONS_BY_PROVIDER_KEY = new Map<
  string,
  readonly MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[]
>([
  ['smtp', freezeMailRuntimeValue([
    SMTP_TRANSPORT_mail_PROVIDER_EXTENSION_DESCRIPTOR
  ])],
  ['imap', freezeMailRuntimeValue([
    IMAP_SYNC_mail_PROVIDER_EXTENSION_DESCRIPTOR
  ])],
]);

export function getMailProviderExtensionCatalog(): readonly MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[] {
  return mail_PROVIDER_EXTENSION_CATALOG;
}

export function getMailProviderExtensionDescriptor(
  extensionKey: string,
): MailProviderExtensionDescriptor<MailKnownProviderExtensionKey> | undefined {
  return mail_PROVIDER_EXTENSION_DESCRIPTOR_BY_KEY.get(extensionKey);
}

export function getMailProviderExtensionsForProvider(
  providerKey: string,
): readonly MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[] {
  return mail_PROVIDER_EXTENSIONS_BY_PROVIDER_KEY.get(providerKey) ?? EMPTY_mail_PROVIDER_EXTENSIONS;
}

export function getMailProviderExtensions(
  extensionKeys: readonly string[],
): readonly MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[] {
  const descriptors: MailProviderExtensionDescriptor<MailKnownProviderExtensionKey>[] = [];

  for (const extensionKey of extensionKeys) {
    const descriptor = mail_PROVIDER_EXTENSION_DESCRIPTOR_BY_KEY.get(extensionKey);
    if (descriptor) {
      descriptors.push(descriptor);
    }
  }

  return freezeMailRuntimeValue(descriptors);
}

export function hasMailProviderExtension(
  extensionKeys: readonly string[],
  extensionKey: string,
): boolean {
  for (const providerExtensionKey of extensionKeys) {
    if (
      providerExtensionKey === extensionKey &&
      mail_PROVIDER_EXTENSION_DESCRIPTOR_BY_KEY.has(extensionKey)
    ) {
      return true;
    }
  }

  return false;
}
